### PR TITLE
Add amber overview doc

### PR DESCRIPTION
**Pull request by dj2**

### CODE REVIEW

**Comment by dj2** [comment]
Ok, I've attempted to translate some of my brain dump into a document.

@sarahM0 I'm sure there are things I'm missing or that aren't explained well, can you please let me know anything you think is missing/poorly explained that would help here? 

@dneto0 anything off-hand you think is needed here? An engine description is probably needed but maybe a doc on its own.

File: docs/amber.md [hunk 1]
```diff
@@ -0,0 +1,138 @@
+# Amber
+
+Amber is a multi-API shader test framework.Graphics bugs can be captured and
```
Comment (dneto0) [inline docs/amber.md]: I insist on at least one space after a sentence-terminating period.  (Normally I use 2 but I won't start a total religous war over it.)

File: docs/amber.md [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+communicated through a scripting interface. This removes the need to program
+to the low level interface when reproducing bugs.
+
+Amber is broken into multiple parts, the applications, the parsing components
```
Comment (dneto0) [inline docs/amber.md]: recommend change the first comma to a colon.

File: docs/amber.md [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+and the script execution.
+
+## Applications
+There are currently two applications, the `samples/amber.cc` amber application
+and the amber integration into the [CTS](https://github.com/KhronosGroup/VK-GL-CTS/tree/master/external/vulkancts/modules/vulkan/amber). These applications are responsible
```
Comment (dneto0) [inline docs/amber.md]: CTS --> "Vulkan Conformance Test Suite"

File: docs/amber.md (hunk 2 above)
Comment (dneto0) [reply docs/amber.md]: Should we change parts --> layers ?   they truly are layers.

File: docs/amber.md [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+for configuring the script execution environment (setting up Vulkan, Dawn or
+another engine), calling into the parsing code to generate a test script and
+then passing that script into the script execution component.
+
+We require the application to configure the execution engine. This allows the
+application to handle loading function pointers, doing configuration and other
+setups as required. There are no hardcoded assumptions in amber as to how you're
```
Comment (dneto0) [inline docs/amber.md]: Always use all-caps when naming "Amber".

File: docs/amber.md [hunk 5]
```diff
@@ -0,0 +1,138 @@
...
+setting up the API to be tested.
+
+This engine configuration is done through the `VulkanEngineConfig` in
+`amber/amber_vulkan.h` and the `DawnEngineConfig` in `amber/amber_dawn.h`.
```
Comment (dneto0) [inline docs/amber.md]: and --> or

(You only have to configure one...)

File: docs/amber.md [hunk 6]
```diff
@@ -0,0 +1,138 @@
...
+
+The sample application does this configuration through the config_helper
+classes. `samples/config_helper_vulkan.*` and `samples/config_helper_dawn.*`.
+For the CTS, the Vulkan engine is already configured and we just set the
+`VulkanEngineConfig` as needed.
+
+Accessing amber itself is done through the `Amber` class in `amber/amber.h`. We
+require the application to parse and execute a script as two separate steps.
+This separation allows for the scripts to be retrieved after a parse command
```
Comment (dneto0) [inline docs/amber.md]: Instead of scripts, do you mean shaders?

File: docs/amber.md [hunk 7]
```diff
@@ -0,0 +1,138 @@
...
+and then pre-compiled and passed into the execution. Passing the compiled
+scripts is optional.
+
+### Delegate
+A delegate object can be provided to Amber, `nullptr` can be provided for the
+delegate. If the delegate is provided and the `LogGraphicsCalls` method returns
+`true` then the vulkan API wrappers will call `Log` for each call into Vulkan.
```
Comment (dneto0) [inline docs/amber.md]: vulkan --> Vulkan

File: docs/amber.md [hunk 8]
```diff
@@ -0,0 +1,138 @@
...
+`VulkanEngineConfig` as needed.
+
+Accessing amber itself is done through the `Amber` class in `amber/amber.h`. We
+require the application to parse and execute a script as two separate steps.
+This separation allows for the scripts to be retrieved after a parse command
+and then pre-compiled and passed into the execution. Passing the compiled
+scripts is optional.
+
+### Delegate
+A delegate object can be provided to Amber, `nullptr` can be provided for the
```
Comment (dneto0) [inline docs/amber.md]: Explain quickly why you'd want a delegate.

"
A delegate object can be provided to Amber, for observing internal operations as the script is executed.
The delegate can be a null pointer, to indicate no observations are requested.
"

File: docs/amber.md [hunk 9]
```diff
@@ -0,0 +1,138 @@
...
+
+### Buffer Extractions
+Amber can be instructed to retrieve the contents of buffers when execution is
+complete. This is done through the `extractions` list in the amber `Options`
+structure. You must set the `buffer_name` for each of the buffers you want to
+extract. When amber completes it will fill out the `width`, `height` and set
+of `Value` objects for that buffer.
+
+### Execution
+There are two methods to execute a parsed script `Execute` and
```
Comment (dneto0) [inline docs/amber.md]: need a colon after "script"

File: docs/amber.md [hunk 10]
```diff
@@ -0,0 +1,138 @@
...
+`ExecuteWithShaderData`. They both accept the `Recipe` and `Options`, the
+`ExecuteWithShaderData` also accepts a map of shader name to data. The data
+is the compiled SPIR-V binary for the shader. This allows you to compile and
+cache the shader if needed.
+
+## Parsing component
+There are currently two parsing backends for Amber.
```
Comment (dneto0) [inline docs/amber.md]: I would rephrase to avoid "backends", since that already has a meaning.

Suggest  "Amber can use scripts written in two dialects:  AmberScript, and VkScript".

File: docs/amber.md [hunk 11]
```diff
@@ -0,0 +1,138 @@
...
+[AmberScript](amber_script.md) and [VkScript](vk_script.md). The `AmberScript`
+parser will be used if the first 7 characters of the input buffer are
```
Comment (dneto0) [inline docs/amber.md]: input buffer --> script

File: docs/amber.md [hunk 12]
```diff
@@ -0,0 +1,138 @@
...
+`#!amber`, otherwise the `VkScript` parser will be used. The parsers both
+generate a `Script` which is our representation of the script file.
+
+### VkScript
+For VkScript we do a bit of work to make the script match AmberScript. A default
+pipeline is generated and all content in the script is considered part of the
+generated pipeline. We generate names for all of the buffers in the file. The
+framebuffer is named `framebuffer`. The generated depth buffer is named
+`depth_buffer`. For each buffer seen we generate a name of `AutoBuf-<num>` where
```
Comment (dneto0) [inline docs/amber.md]: "For each buffer seen we generate ..." --> "For other buffers, we generate ..."

File: docs/amber.md [hunk 13]
```diff
@@ -0,0 +1,138 @@
...
+There are currently two parsing backends for Amber.
+[AmberScript](amber_script.md) and [VkScript](vk_script.md). The `AmberScript`
+parser will be used if the first 7 characters of the input buffer are
+`#!amber`, otherwise the `VkScript` parser will be used. The parsers both
+generate a `Script` which is our representation of the script file.
+
+### VkScript
+For VkScript we do a bit of work to make the script match AmberScript. A default
+pipeline is generated and all content in the script is considered part of the
+generated pipeline. We generate names for all of the buffers in the file. The
```
Comment (dneto0) [inline docs/amber.md]: "An Amber buffer corresponds to either a buffer resource or an image resource in the backend engine's API.  We generate a unique name for each buffer in the script."

File: docs/amber.md [hunk 14]
```diff
@@ -0,0 +1,138 @@
...
+
+The VkScript parser is broken into three major chunks. The `Parser`,
+`SectionParser` and `CommandParser`. The `Parser` is the overall driver for the
+parser. The `SectionParser` breaks the input file into the overall chunks (each
+of the sections separated by the \[blocks]). The `CommandParser` parses the
+`[test]` section specifically. For other sections they're parsed directly in
+the `Parser` object.
+
+### Parsed Representation
+The `Script` object is the owner for all of our data. Other objects hold
```
Comment (dneto0) [inline docs/amber.md]: "The `Script` object owns all script data."

(Surely it doesn't own _all_ the data.)

File: docs/amber.md [hunk 15]
```diff
@@ -0,0 +1,138 @@
...
+Type           |                  |  Binding              |                       |
+Compiled Shader|                  |  Attachment Location  |                       |
+               |                  |                       |                       |
+               +------------------------------------------+                       |
+                                  |                                               |
+                                  +-----------------------------------------------+
+                                                        BufferCommand
+```
+
+A `Script` is made up for shaders, pipelines, buffers and commands. Pipelines
```
Comment (dneto0) [inline docs/amber.md]: Oxford commas please, and tweaking words:

"A `Script` is contains shaders, buffers, pipelines, and commands."

File: docs/amber.md [hunk 16]
```diff
@@ -0,0 +1,138 @@
...
+`depth_buffer`. For each buffer seen we generate a name of `AutoBuf-<num>` where
+the number is the current number of buffers seen counting from 0.
+
+The VkScript parser is broken into three major chunks. The `Parser`,
+`SectionParser` and `CommandParser`. The `Parser` is the overall driver for the
+parser. The `SectionParser` breaks the input file into the overall chunks (each
+of the sections separated by the \[blocks]). The `CommandParser` parses the
+`[test]` section specifically. For other sections they're parsed directly in
+the `Parser` object.
+
```
Comment (dneto0) [inline docs/amber.md]: You described how VkScript is parsed.  But you didn't mention how AmberScript is parsed.  Maybe it's enough to say an AmberScript script is mapped pretty much directly to the parsed representation described in the next section.

File: docs/amber.md [hunk 17]
```diff
@@ -0,0 +1,138 @@
...
+Compiled Shader|                  |  Attachment Location  |                       |
+               |                  |                       |                       |
+               +------------------------------------------+                       |
+                                  |                                               |
+                                  +-----------------------------------------------+
+                                                        BufferCommand
+```
+
+A `Script` is made up for shaders, pipelines, buffers and commands. Pipelines
+contain shaders and buffers. The buffers hold both image and storage buffers.
```
Comment (dneto0) [inline docs/amber.md]: Instead of "The buffers hold both image and storage buffers", suggest:  "An Amber buffer corresponds to either a buffer resource or an image resource in the backend engine's API. "

(Except that sentence only belongs in one place in this doc, and I already suggested it above.)

File: docs/amber.md (hunk 13 above)
Comment (dneto0) [reply docs/amber.md]: (hmm. but see below, where I suggested a different place for that first sentence.)

File: docs/amber.md [hunk 18]
```diff
@@ -0,0 +1,138 @@
...
+It is assumed that the `Buffer` object has the current value for that buffer at
```
Comment (dneto0) [inline docs/amber.md]: "Script execution assumes that after executing a command, each `Buffer` object has a reference to the latest data for that buffer or image mapped into host (CPU) memory."

File: docs/amber.md [hunk 19]
```diff
@@ -0,0 +1,138 @@
...
+colour attachments will have their `Attachment Location` while data buffers
+will have a `Descriptor Set` and `Binding` provided.
+
+## Execution
+When the script is executed the pipeline shaders will be compiled, if not
+provided through the shader map. This will fill in the `Compiled Shader` data
+in the each pipeline shader object. The `CreatePipeline` call will then be
+executed for a given pipeline.
+
+With the pipeline created on the backed each of the `Command` objects will
```
Comment (dneto0) [inline docs/amber.md]: Grammar?  I can't parse this, especially around "the backed"

File: docs/amber.md [hunk 20]
```diff
@@ -0,0 +1,138 @@
...
+be executed. Each `Command` knows the amber `Pipeline` associated with it, that
+`Pipeline` pointer is provided in the command execution and can be used to
+lookup the engine specific representation of a pipeline.
```
Comment (dneto0) [inline docs/amber.md]: "engine-specific"

File: docs/amber.md [hunk 21]
```diff
@@ -0,0 +1,138 @@
...
+
+When the `Probe` and `ProbeSSBO` commands are encountered they are not passed
+to the engine but sent to the `Verifier`. This will validate that the data
+in the requested buffer matches the test data. As mentioned above, this assumes
```
Comment (dneto0) [inline docs/amber.md]: requested --> specified

File: docs/amber.md (hunk 1 above)
Comment (dj2) [reply docs/amber.md]: Done. I like one, you're wasting soooo many bytes by using 2. Won't you think of the spinning disks?

File: docs/amber.md (hunk 2 above)
Comment (dj2) [reply docs/amber.md]: done. layers is a better name, thanks.

File: docs/amber.md (hunk 3 above)
Comment (dj2) [reply docs/amber.md]: done. Made it "Vulkan Conformance Test Suite 'CTS'" as I use CTS later.

File: docs/amber.md (hunk 4 above)
Comment (dj2) [reply docs/amber.md]: done

File: docs/amber.md (hunk 5 above)
Comment (dj2) [reply docs/amber.md]: done

File: docs/amber.md (hunk 6 above)
Comment (dj2) [reply docs/amber.md]: Yes, yes I do.

File: docs/amber.md (hunk 8 above)
Comment (dj2) [reply docs/amber.md]: done

File: docs/amber.md (hunk 7 above)
Comment (dj2) [reply docs/amber.md]: done

File: docs/amber.md (hunk 9 above)
Comment (dj2) [reply docs/amber.md]: done

File: docs/amber.md (hunk 10 above)
Comment (dj2) [reply docs/amber.md]: done

File: docs/amber.md (hunk 11 above)
Comment (dj2) [reply docs/amber.md]: done

File: docs/amber.md (hunk 12 above)
Comment (dj2) [reply docs/amber.md]: done

File: docs/amber.md (hunk 14 above)
Comment (dj2) [reply docs/amber.md]: Better?

The `Script` object owns all of the pipelines, buffers, shaders, and command
objects. Other objects hold pointers but the script holds the `unique_ptr` for
these objects.

File: docs/amber.md (hunk 16 above)
Comment (dj2) [reply docs/amber.md]: done

File: docs/amber.md (hunk 15 above)
Comment (dj2) [reply docs/amber.md]: done

File: docs/amber.md (hunk 13 above)
Comment (dj2) [reply docs/amber.md]: added it below where I think it makes more sense.

File: docs/amber.md (hunk 17 above)
Comment (dj2) [reply docs/amber.md]: added here

File: docs/amber.md (hunk 18 above)
Comment (dj2) [reply docs/amber.md]: done. Although, it's not CPU memory, the buffer itself holds a chunk of memory the values have to be copied into.

File: docs/amber.md (hunk 19 above)
Comment (dj2) [reply docs/amber.md]: done.

File: docs/amber.md (hunk 20 above)
Comment (dj2) [reply docs/amber.md]: done

File: docs/amber.md (hunk 21 above)
Comment (dj2) [reply docs/amber.md]: done

File: docs/amber.md [hunk 22]
```diff
@@ -0,0 +1,150 @@
+# Amber
+
+Amber is a multi-API shader test framework. Graphics bugs can be captured and
```
[the remark this answers is not in the document]
Comment (dj2) [reply docs/amber.md]: Good catch. Fixed.

File: docs/amber.md [hunk 23]
```diff
@@ -0,0 +1,150 @@
...
+observations are requested.
+
+If the delegate is provided and the `LogGraphicsCalls` method returns
+`true` then the Vulkan API wrappers will call `Log` for each call into Vulkan.
+If the `LogGraphicsCallsTime` also returns `true` then timings for those
+Vulkan calls will also be recorded. The timestamps are retrieved from the
+`GetTimestampNS` callback.
+
+### Buffer Extractions
+Amber can be instructed to retrieve the contents of buffers when execution is
```
[the remark this answers is not in the document]
Comment (dj2) [reply docs/amber.md]: That sounds like something I should write down in another document. There are two concepts, buffers and commandbuffers which can be confusing. For the buffers, we're talking things that just contain bits. Those bits could be images, compute results or input data. For commandbuffers their engine specific things for sending commands to the devices.

The data buffers you'll see colour attachments, depth/stencil attachments which are related to drawing things and then storage, uniform and push constants which are about data. Push constants are small but fast. I believe uniform is global for all shader invocations, I believe it doesn't change. Storage is read/write.
